Additional API changes (C4-363)

netsettler opened the pull request:
This branch-to-branch PR adds fixes to the PR that is already approved. They were not trivial, so I made to this branch over vacation to capture them without revisiting other work. This fixes issues from code review and bugs that were found in more thorough testing.

* Newly uses `dcicutils 1.8.1` in lieu of removed file `qa_utils_subset.py`, which had some duplicated functionality. There were some compatibility issues that had to be addressed because I changed the interfaces when installing them in `qa_utils`, so a few small changes to calls had to be made. e.g.,
  * `in_date_range` became `in_date_interval`
  * Use new `classify_server_url` to figure out what kind of server the referer url is.
* Created a prd/stg distinction so this functionality is more easily tested.
  * `CALENDAR_DATA_URL` became `CALENDAR_DATA_URL_PRD` and `CALENDAR_DATA_URL_STG`.
  * The appropriate calendar data URL is chosen based on whether `PRD_ENDPOINT_PATH` (`/4dn-status`) or `STG_ENDPOINT_PATH` (`/4dn-status-staged`) is used, so that each of the production and staged repos has different data to allow for easy testing as a lambda function.
* Data files no longer have to contain timezones. They will be parsed as HMS-relative (US/Eastern) and so will get the appropriate daylight time best if a TZ is not given. It is OK (but error prone) to still use a -0500 or -0400 offset. Better to just omit it.
* Added a `try` around the main handler,  yielding better error messaging if an error occurs.

willronchetti commented on functions/4dn-status/src/lambda_function.py:
```diff
@@ -1,14 +1,16 @@
+# import datetime
 import html
-import http
+# import http
 import io
 import json
-import pytz
-import re
+# import pytz
+# import re
```
Purpose for commenting out vs. removing?

netsettler replied:
Just the usual making sure things work first. I'll remove them. They were implementation-level support for things that got removed in the `dcicutils` switchover, I think.

netsettler replied:
Fixed.

willronchetti commented on functions/4dn-status/src/lambda_function.py:
```diff
@@ -33,18 +35,26 @@
 
 
 def convert_to_html(data, environment):
+    ignored(environment)  # TODO: Should this be ignored?
```
Good question. When you test this, what is in it?

netsettler replied:
It's (approximately) an ffenv. In particular, it's the bucket env, so for staging/prod it's been canonicalized to `fourfront-webprod` on the `fourfront` app and to `fourfront-cgap` for `cgap` app.  I don't know that the way in which this conversion occurs needs to care about that, but it could. I left the dataflow just so it's easy to find if it turns out later we care.  Originally I thought it might be used to do filtering of items, but that's now done farther upstream. Even without filtering, though, it _could_ still be used to pull up a different color or icon or something, though. Just probably overkill so I'm not expecting it.

netsettler replied:
And in fact the next comment you made is about the logo thing. So I'll try to connect it to that.

netsettler replied:
Fixed by making it control the icon and header text to be CGAP- or Fourfront-specific.

willronchetti commented on functions/4dn-status/src/lambda_function.py:
```diff
@@ -88,7 +98,9 @@ def convert_to_html(data, environment):
     <tr>
-     <td valign="middle"><img src="https://4dnucleome.org/Assets/images/4dn-logo_1.png" class="logo" alt="4D Nucleome" /></td>
+     <td valign="middle">
+      <img src="https://4dnucleome.org/Assets/images/4dn-logo_1.png" class="logo" alt="4D Nucleome" />
```
If I deployed this for CGAP I would get a 4DN logo here, so I think you might need to pass an argument or compute this URL.

netsettler replied:
Ah, yes, well, this is why the environment argument would get used.

netsettler replied:
Fixed.

willronchetti reviewed (approved):
Approving with some minor comments

netsettler commented:
In the most recent round of changes, I added the ability to pass `now=<date>` to test against a given date.

I also added the ability to pass `debug=True` (in any case for 'true') in order to get additional debugging data when the filtering is doing weird stuff. I also made it return a `"problems"` field in the result when there is some weird problem in the obtaining of the data file or the filtering because in manual testing I noticed that if the file doesn't get found, it's quite mysterious. That's not a bug, but it's easier to debug if some kind of error gets returned, even though the calendar it returns tries to be robust in the face of such an error (so the default calendar is returned, but the problems element says why). There is a test case for this.

The reason why this is vexing is that if you use the AWS console to upload a new calendar, you have to manually re-update the file to be world-readable because it won't inherit the previous value. In the interim, it will not be found. We could use a better permission scheme, and may have to ultimately do that to satisfy other hospitals, but really there's nothing wrong with our status info being world-readable.

While I was in there, I also changed the format of the data file to use the field name `"calendar"` rather than `"events"` so it doesn't get confused with AWS events. I likewise changed the data file to be called `calendar.json` (or `calendar-staged.json`).

And I changed the result headers in the HTML case to be dependent on the environment type, so that you get a CGAP or Fourfront logo and heading as appropriate.

netsettler commented:
I don't think these changes will be controversial and I want to get this merged. I added some tests, but also did some manual testing. Things seem working, so I'm going to merge this down and deploy it down into the main PR. If there's no objection in the moment, I'll merge that PR and be done.

willronchetti commented on functions/4dn-status/src/lambda_function.py:
```diff
@@ -47,26 +43,42 @@ def get_calendar_data(staged=False):
+        return data
+
+
+CGAP_LOGO_URL = "https://cgap.hms.harvard.edu/static/img/exported-logo.svg"
+FF_LOGO_URL = "https://data.4dnucleome.org/static/img/4dn_logo.svg"
```
We might consider serving these from somewhere else, but not something you can deal with now. Just wanted to make note of it.
